### PR TITLE
fix(loss): reject InfoNCE directions without candidates to score

### PR DESCRIPTION
**Pull request by kabirvashisht4-glitch**
# What does this PR do ?

Rejects the InfoNCE configurations whose loss is mathematically forced to `0.0`,
instead of training on them silently.

`infonce_loss` / `infonce_distill_loss` accept `direction="d2q"` or
`"symmetric"` together with `use_in_batch_negatives=False`. Hard negatives are
documents mined *per query*, so they are candidates for the query-to-document
direction only, and `_one_direction` correctly passes `None` for the reverse
direction. With in-batch negatives also switched off, that direction is left
with a single candidate column:

```python
pos = (query * doc).sum(dim=-1, keepdim=True)   # [B, 1]
logits = pos
target = torch.zeros(batch, device=device, dtype=torch.long)
...
return F.cross_entropy(logits, target)          # one class -> log(1) == 0
```

The existing guard only caught the case where *no* negatives were passed at all,
so it never fired:

```python
if not use_in_batch_negatives and not has_hard_negs:
    raise ValueError("infonce_loss: no negatives provided")
```

Measured on `main` @ `b1f4031d`, `B=4`, `K=3`, CPU:

| direction | `use_in_batch_negatives` | loss | sum abs grad (queries) | sum abs grad (hard negs) |
|---|---|---:|---:|---:|
| `q2d` | true | 8.986291 | 25.4025 | 4.2906 |
| `q2d` | false | 5.643632 | 17.9486 | 20.5227 |
| `d2q` | true | 5.800508 | 20.0058 | — |
| `d2q` | false | **0.000000** | **0.0000** | — |
| `symmetric` | true | 7.393400 | 20.2241 | 2.1453 |
| `symmetric` | false | 2.821816 | 8.9743 | 10.2613 |

`2.821816 == 5.643632 / 2` exactly — the `d2q` half of `"symmetric"` contributes
literally nothing, so a config asking for a symmetric objective silently gets a
halved one-directional one. `infonce_distill_loss` degenerates identically: the
single-candidate softmax makes the student and the detached teacher
distributions both one-hot, so KL/CE/MSE over them is `0.0` with zero student
gradients.

This is reachable from YAML: `RetrievalDistillBiEncoderRecipe` builds both
modules from `self.cfg.get("infonce_loss")` / `self.cfg.get("infonce_distill_loss")`,
and `direction` and `use_in_batch_negatives` are plain keys on that block (see
`examples/retrieval/distillation/nemotron3_embed_1b_distill.yaml`). The shipped
example is safe; flipping those two keys is not.

## Why reject rather than feed the negatives into `d2q`

Reusing `hard_negatives` for the reverse direction would be numerically wrong,
not just unhelpful: they are *document* embeddings, and the document-to-query
direction needs query-side candidates. The `None` in `_one_direction(d, q, None)`
is correct; the missing piece is that the caller was never told the resulting
candidate set is empty. So the guard is extended instead, in the same place and
the same style as the existing one, and the failure is raised before any work is
done. This matches the precedent already in the recipe:

```python
# Guard against a configured intermediate loss whose layer_pairs are empty
# while the weight is active: hooks would fire but the loss would silently
# short-circuit to zero (contributing no gradient).
```

No behavior changes for any configuration that was producing a real loss.

# Changelog

- `infonce_loss` and `infonce_distill_loss` now raise `ValueError` when
  `use_in_batch_negatives=False` is combined with `direction="d2q"` or
  `"symmetric"`, explaining that the document-to-query direction can only draw
  candidates from the batch and naming both ways out.
- Both functions gained full Google-style docstrings recording the tensor
  layouts (`[batch, hidden]`, `[batch, negatives, hidden]`, `[batch, negatives]`
  masks), the per-query/document-side semantics of `hard_negatives` that make
  the direction restriction necessary, and the `Raises` contract.
- `tests/unit_tests/loss/test_infonce.py`: added the rejection tests for both
  functions and for the `InfoNCELoss` module (the config-driven entry point),
  plus parametrized guards asserting every *accepted* `(direction,
  use_in_batch_negatives)` combination yields a strictly positive loss and
  non-zero input gradients — the invariant the bug violated.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

The only documentation of this contract was the functions' one-line docstrings,
which are expanded here. No user-facing docs or example configs reference the
rejected combination, so nothing else needed updating.

Verified locally: `pytest tests/unit_tests/loss/test_infonce.py` -> 62 passed.
The 5 new rejection tests fail on `main` (`DID NOT RAISE ValueError`) and pass
here. `ruff format --check` / `ruff check` clean. CPU only, no GPU needed.

# Additional Information

- Related to #3869

### CODE REVIEW

**Review by yuhezhang-ai — APPROVED** [review]
This is useful housekeeping for an uncommon configuration. The supplied example is unaffected, and no affected training run is reported here. Rejecting d2q/symmetric with in-batch negatives disabled makes the configuration contract explicit and prevents silently dropping the reverse-direction loss.

Validated the focused loss suite locally on CPU: 62 tests passed.

**Comment by yuhezhang-ai** [comment]
/ok to test [0583390](https://github.com/NVIDIA-NeMo/Automodel/pull/3870/commits/0583390d86ef7624f5e4d469f37631a791a02428)
